### ISSUE TITLE
Slow response time in Azure Monitor metrics resource selector

### ISSUE THREAD

**Issue by kayodeprinceMS** [grafana/grafana#92887]
### What happened?

With the introduction of current-user auth, enterprise level Azure customers will have access to dozens or hundreds of subscriptions, resource groups and resources. The current approach to populate this hierarchical list in the resource selector for Azure Monitor metrics can result in response times of 2 minutes or more. And leads to a poor UX for simply finding and selecting the right set of resources when authoring a panel.

On the cold cache case, expanding a single subscription in Resource Selector issues API calls for all the resources across all subscriptions (not just the one being selected).

https://github.com/user-attachments/assets/6dca1692-6054-4df1-903f-79438f1ca569

I've taken a quick look at the difference in approaches and performance for populating the resource group list based on selecting a single subscription. The Grafana resource selector take approximately 1 min 50 sec compared to Portal's 2-3 sec response time (with cold browser cache).

This slow response time also applies in the case of searches in Metrics Resource selector which searches across all subs.

### What did you expect to happen?

Expected response time for load resource groups within a single subscription to be a few seconds, 1-3 seconds at most.

### Did this work before?

No

### How do we reproduce it?

1. Connect Azure Monitor datasource to Azure using an auth principal that has read access to many subscriptions e.g. 50+subs
2. Clear browser cache
3. Create new panel with Azure Monitor metrics service
4. Click on Resource Selector

### Environment (with versions)?

Grafana: 11.1
OS: Windows
Browser: Chrome and Edge

### Grafana platform?

Other

### Datasource(s)?

Azure Monitor

**Comment by aangelisc** [comment]
Hi @kayodeprinceMS, following up with a potential implementation:

We're considering switching from requesting the namespaces on an individual subscription to level to only requesting the namespaces from the following regions and then combining them.
 
The regions we would use are: `EastUS`, `West Europe`, and `Japan East`.
 
I’d like to confirm if this API request is subscription agnostic. E.g.
 
We have SubscriptionA and SubscriptionB
SubscriptionA which (for example) doesn’t have the same resource providers registered as SubscriptionB (e.g. Microsoft.Storage is registered in B but not A)
 
Will the request against SubscriptionA still return Microsoft.Storage in the namespace list? If so then I believe this is fine and we can proceed with this method which will lead to 3 static requests being made based on the first available subscription rather than a request being issued per subscription. If this is not the case then we’ll have to rethink how this works.
 
The reason this functionality is essential is to support searching the resources available to a user. If the above is not true then we can support the resource picker appropriately and make the request for namespaces only when a subscription is expanded. However, this will definitely break the search functionality as some resources may not be searchable.

**Comment by kayodeprinceMS** [comment]
Hi @aangelisc thanks for this update. I've started testing the UI with the new resource selector behavior. It works as described and as expected in most cases. 

However, there is a case where the datasource identity (current user) may not have reader permissions on the 1st subscription in the list and only has permissions on some subset of resources within the subscription. In this case, the metrics namespaces API still returns 200 OK but only includes an empty array instead of the full list namespaces.  

![Image](https://github.com/user-attachments/assets/d0b49e72-9d30-4800-b50f-72dd49416355)

When this occurs, the plugin throws an error and does not retrieve the list of namespaces from any other subscription.

![Image](https://github.com/user-attachments/assets/2831075f-b09a-4c9e-ad00-aafe81a49adc)

Prior to introducing user-based auth and the new resource selector behavior, if the datasource identity (MI or App Reg) did not have read access on the entire subscription, expanding the subscription would not show any resource groups or resources and the user would need to manually input the resource details.

![Image](https://github.com/user-attachments/assets/dc1780d1-6cb3-4768-86d6-af38c34b04d2)

One possible approach is to retain the new behavior for querying metric namespaces API using 3 regions but to do so for each sub independently in parallel or on demand per user expansion to keep the performance improvements.

I'll let @avidhanju chime in with specific recommendations if there are any from our team.

**Comment by aangelisc** [comment]
Hi @kayodeprinceMS,

At the moment the expectation in Grafana is that the user will have reader permissions on all subscriptions they're able to query.

We could fallback to the default hardcoded list in Grafana in this case and display an error to the user.

I'm not keen on implementing something that will query each sub independently in parallel as this will not scale well. Additionally, querying on-demand will mean that a user will be unable to make use of the search functionality if the first subscription does not return any namespaces.

**Comment by kayodeprinceMS** [comment]
Understood. Ack on the scaling and search concerns. It makes sense to ensure that the solution accounts for both of these cases. I'm onboard with different fallback behavior.

However, "the expectation in Grafana is that the user will have reader permissions on all subscriptions they're able to query" may be too broad to hold as the default position. Users can often have access to resource groups and individual resources without having read access on the entire subscription. We should still support the ability to navigate and select these resource groups and resources based on the current-user access.

As a reference point, one of the most customer reported issues on the Azure Monitor out-of-the-box dashboards is that the variable selection chain to choose the resource only works if the MI/App registration configured in the data source has subscription level permissions.

**Comment by aangelisc** [comment]
We can discuss the permissions requirement. At the moment, IIRC, the reason for this limitation is due to the limitations of the metric namespaces API. We can only query the API at the subscription level as it's not possible for us to determine what resources a user may have access to.

Supporting navigating resource groups and resources when the user doesn't have access to the subscription (or resource group) would be a separate feature request.

**Comment by aangelisc** [comment]
We will implement the fallback behaviour to the existing hardcoded list. Once you've created the new issue please link it here 😊

**Comment by kayodeprinceMS** [comment]
Adding link to new issue https://github.com/grafana/grafana/issues/94500 to request support for the ability to navigate and select resource groups and resources when the principal does not have subscription level read access.
